protocol/contracts/token/ZaiV2.sol

```diff
@@ -0,0 +1,71 @@
+/*
+    Copyright 2020 Zero Collateral Devs, standing on the shoulders of the Empty Set Squad <zaifinance@protonmail.com>
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+*/
+
+pragma solidity ^0.5.17;
+pragma experimental ABIEncoderV2;
+
+import "./Dollar.sol";
+import "./IDollar.sol";
+
+contract ZaiV2 is Dollar {
+    // 100:1 ZAI for ZAIv2
+    uint256 public constant SWAP_EXCHANGE_RATE = 100;
+
+    // Total supply of ZAI at the time of reset. Cannot swap more than that
+    uint256 public constant MAX_ZAI_SWAPPABLE = 395407172763503943977588203;
+
+    // Track zai that has been swapped
+    uint256 public oldZaiSwapped = 0;
+
+    // Original ZAI token address must be this on mainnet.
+    // address(0x9d1233cc46795E94029fDA81aAaDc1455D510f15);
+    // Only settable in the constructor for testing
+    IDollar internal _oldZai;
+
+    event BurnAndSwap(
+        address indexed operator,
+        address indexed from,
+        uint256 burnAmount,
+        uint256 mintAmount
+    );
+
+    constructor(address oldZaiAddress) public Dollar() {
+        _oldZai = IDollar(oldZaiAddress);
+    }
+
+    // Approve the amount of ZAI for the ZAIv2 contract to burn and swap, then
+    // call burnAndSwap for the address.
+    function burnAndSwap(address from) public {
+        uint256 burnAmount = _oldZai.allowance(from, address(this));
+        require(burnAmount > 0, "gotta burn if you wanna earn...");
+
+        oldZaiSwapped = oldZaiSwapped.add(burnAmount);
+        require(
+            oldZaiSwapped <= MAX_ZAI_SWAPPABLE,
+            "where did all that ZAI come from..."
+        );
```

Comment:
@redacted007 I'm just a solidity beginner, but wondering if you don't have an epoch or time limit on the burnAndSwap function, if that leaves the door open for a dangerous arbitrage opportunity on old zai. 

So, if zaiv2 takes off, and zai still exists and is sale-able, won't this leave open the ability to buy old zai, then swap into zaiv2 for a quick sale? Seems like if so, that either caps the future value of zaiv2 or makes a bug similar in effect to old zai.

Reply:
Long term the price of ZAI is supposed to equal DAI. We want people to arb the supply of v1 to v2 (which is why it's 100:1) The total that's allowed to swap is capped at the total supply of v1 (3.96m in v2 terms.) In relative terms, as the protocol grows, that 3.96m will represent a small portion of the overall value of the network. It's not a "dangerous" arb, it's going to work as intended!

That arb is the best way to continue the vision of ZAIon!

Reply:
Thanks for the reply! I see...the cap really is the supply of ZAI v1, and we don't care if somebody were to buy most of that supply and arb because it's a drop in the bucket of ZAI v2. I ran a couple price vs supply vs arb scenarios and it looks like all is right. 
Thanks again for all your work on this.